src/hotspot/cpu/aarch64/assembler_aarch64.hpp

```diff
@@ -2684,12 +2684,28 @@ void mvnw(Register Rd, Register Rm,
      *   01xx xxx       2S/4S,  shift = UInt(immh:immb) - 32            \
      *   1xxx xxx       1D/2D,  shift = UInt(immh:immb) - 64            \
      *   (1D is RESERVED)                                               \
+     *                                                                  \
+     * Shift is zero?                                                   \
+     *   Not accumulte:  generate an 'orr', actually a 'mov'.           \
+     *   Need accumulte: generate an 'add'.                             \
      */                                                                 \
     assert((1 << ((T>>1)+3)) > shift, "Invalid Shift value");           \
-    int cVal = (1 << (((T >> 1) + 3) + (isSHR ? 1 : 0)));               \
-    int encodedShift = isSHR ? cVal - shift : cVal + shift;             \
-    f(0, 31), f(T & 1, 30), f(opc, 29), f(0b011110, 28, 23),            \
-    f(encodedShift, 22, 16); f(opc2, 15, 10), rf(Vn, 5), rf(Vd, 0);     \
+    if (shift == 0) {                                                   \
+      bool accumulate = ((opc2 & 0b100) != 0);                          \
+      if (accumulate && isSHR) {                                        \
+        f(0, 31), f(T & 1, 30), f(0b001110, 29, 24);                    \
+        f(T >> 1, 23, 22), f(1, 21), rf(Vn, 16);                        \
+        f(0b100001, 15, 10), rf(Vd, 5), rf(Vd, 0);                      \
+      } else {                                                          \
+        f(0, 31), f(T & 1, 30), f(0b001110101, 29, 21);                 \
+        rf(Vn, 16), f(0b000111, 15, 10), rf(Vn, 5), rf(Vd, 0);          \
+      }                                                                 \
+    } else {                                                            \
+      int cVal = (1 << (((T >> 1) + 3) + (isSHR ? 1 : 0)));             \
+      int encodedShift = isSHR ? cVal - shift : cVal + shift;           \
+      f(0, 31), f(T & 1, 30), f(opc, 29), f(0b011110, 28, 23),          \
+      f(encodedShift, 22, 16); f(opc2, 15, 10), rf(Vn, 5), rf(Vd, 0);   \
+    }                                                                   \
   }
```

Comment:
Is this correct, according to the definition in the Architecture Reference Manual? It doesn't look like it to me. Assembler methods should generate bit patterns exactly as defined in the Manual. This logic should be in a MacroAssembler method.

Reply:
Hi, I moved the logic into MacroAssembler.
The assert is kept to make sure that we would never pass a zero right shift to assemlber.